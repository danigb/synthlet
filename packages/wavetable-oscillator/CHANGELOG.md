# @synthlet/wavetable-oscillator

## 0.1.0

- Initial implementation
