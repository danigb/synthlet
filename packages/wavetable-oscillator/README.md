# @synthlet/wavetable-oscillator

> A morphing wavetable oscillator module for [synthlet](https://github.com/danigb/synthlet)

Wavetables are from [WaveEdit Online](https://waveeditonline.com/)

## Install

```bash
npm i @synthlet/wavetable-oscillator
```
