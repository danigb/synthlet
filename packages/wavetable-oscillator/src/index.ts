import { PROCESSOR } from "./processor";
import { Wavetable, WavetableLoader } from "./wavetable-loader";

export type ProcessorOptions = {};

export type WavetableParams = {
  baseFrequency: number;
  frequency: number;
  morphFrequency: number;
};

export type WavetableWorkletNode = AudioWorkletNode & {
  baseFrequency: AudioParam;
  frequency: AudioParam;
  morphFrequency: AudioParam;
  setWavetable: (wavetable: Float32Array, wavetableLength: number) => void;
};

const PARAM_NAMES = ["baseFrequency", "frequency", "morphFrequency"] as const;

export function getProcessorName() {
  return "WavetableWorkletProcessor"; // Can't import from worklet because globals
}

export function getWorkletUrl() {
  const blob = new Blob([PROCESSOR], { type: "application/javascript" });
  return URL.createObjectURL(blob);
}

/**
 * Load a wavetable
 * @param url
 * @param wavetableLength
 */
export function loadWavetable(
  nameOrUrl: string,
  wavetableLength = 256
): Promise<Wavetable> {
  const url = nameOrUrl.startsWith("http")
    ? nameOrUrl
    : `https://smpldsnds.github.io/wavedit-online/samples/${nameOrUrl.toUpperCase()}.WAV`;
  return new WavetableLoader(url, wavetableLength).onLoad();
}

function isSupported(audioContext: AudioContext): boolean {
  return (
    audioContext.audioWorklet &&
    typeof audioContext.audioWorklet.addModule === "function"
  );
}

function isRegistered(audioContext: AudioContext): boolean {
  return (audioContext.audioWorklet as any).__SYNTHLET_ADSR_REGISTERED__;
}

function register(audioContext: AudioContext): Promise<void> {
  (audioContext.audioWorklet as any).__SYNTHLET_ADSR_REGISTERED__ = true;
  return audioContext.audioWorklet.addModule(getWorkletUrl());
}

/**
 * Register the AudioWorklet processor in the AudioContext.
 * No matter how many times is called, it will register only once.
 *
 * @param audioContext
 * @returns A promise that resolves when the processor is registered
 */
export function registerWavetableOscillatorWorkletOnce(
  audioContext: AudioContext
): Promise<void> {
  if (!isSupported(audioContext)) throw Error("AudioWorklet not supported");
  return isRegistered(audioContext)
    ? Promise.resolve()
    : register(audioContext);
}

/**
 * Creates a Wavetable Oscillator node.
 */
export function createWavetableOscillatorNode() {}

function createWorkletNode(
  audioContext: AudioContext,
  processorOptions: ProcessorOptions,
  params: Partial<WavetableParams> = {}
): AudioWorkletNode {
  const node = new AudioWorkletNode(audioContext, getProcessorName(), {
    numberOfInputs: 1,
    numberOfOutputs: 1,
    processorOptions,
  }) as WavetableWorkletNode;

  for (const paramName of PARAM_NAMES) {
    const param = node.parameters.get(paramName)!;
    const value = params[paramName as keyof WavetableParams];
    if (typeof value === "number") param.value = value;
    node[paramName] = param;
  }

  node.setWavetable = (wavetable, wavetableLength) => {
    node.port.postMessage({
      type: "WAVETABLE",
      wavetable,
      wavetableLength,
    });
  };

  let _disconnect = node.disconnect.bind(node);
  node.disconnect = (param?, output?, input?) => {
    node.port.postMessage({ type: "DISCONNECT" });
    // @ts-ignore
    return _disconnect(param, output, input);
  };
  return node;
}
