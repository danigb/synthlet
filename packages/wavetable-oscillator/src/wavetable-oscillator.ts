type Inputs = {
  baseFrequency: number[];
  frequency: number[];
  morphFrequency: number[];
};

export function WavetableOscillator(sampleRate: number) {
  let $baseFrequency = 220;
  let $frequency = 440;
  let $morphFrequency = 0.005;
  let $wavetable = new Float32Array(0);

  let len = 0;
  let planes = 0;
  let planeA = 0;
  let planeB = 0;
  let offset = 0;
  let inc = 2;
  let morphPhase = Phasor(sampleRate);
  let morphChange = Trigger();

  function read(inputs: Inputs) {
    if (
      inputs.frequency[0] !== $frequency ||
      inputs.baseFrequency[0] !== $baseFrequency
    ) {
      $baseFrequency = inputs.baseFrequency[0];
      $frequency = inputs.frequency[0];
      inc = $frequency / $baseFrequency;
    }
    $morphFrequency = inputs.morphFrequency[0];
  }

  function set(wavetable: Float32Array, length: number) {
    $wavetable = wavetable;
    len = Math.min(length, wavetable.length);
    planes = Math.floor(wavetable.length / len);
    planeA = 0;
    planeB = (planeA + 1) % planes;
  }

  function agen(output: Float32Array, inputs: Inputs) {
    if (len === 0 || planes === 0) {
      output.fill(0);
      return;
    }

    read(inputs);
    for (let i = 0; i < output.length; i++) {
      let morph = morphPhase($morphFrequency);
      const a = interpolateLinear2d($wavetable, len, planeA, offset);
      if (planeB !== planeA) {
        const b = interpolateLinear2d($wavetable, len, planeB, offset);
        output[i] = (1 - morph) * a + morph * b;
      } else {
        output[i] = a;
      }
      offset += inc;
      if (offset >= len) offset -= len;
      if (offset < 0) offset += len;

      if (morphChange(morph)) {
        // TODO: on plane change a click is audible (fix it)
        planeA = planeB;
        planeB = (planeB + 1) % planes;
      }
    }
  }

  return { agen, set };
}

/**
 * Linear interpolation for a 2d buffer
 */
function interpolateLinear2d(
  buffer: Float32Array,
  len: number,
  plane: number,
  offset: number
) {
  const index = Math.floor(offset);
  const frac = offset - index;
  const next = (index + 1) % len;
  const y1 = buffer[plane * len + index];
  const y2 = buffer[plane * len + next];
  const y = y1 + (y2 - y1) * frac;
  return y;
}

/**
 * A Phasor is a signal generator that produces a sawtooth wave with a range of -1 to 1 in a specified frequency.
 */
function Phasor(sampleRate: number) {
  let isr = 1 / sampleRate;
  let phase = 0;

  return (frequency: number) => {
    phase += frequency * isr;
    while (phase >= 1.0) phase -= 1.0;
    while (phase < 0.0) phase += 1.0;
    return phase;
  };
}

/**
 * Convert a ramp into a trigger
 * @returns
 */
function Trigger() {
  let prev = 0;
  let prevWasTrigger = false;

  return function trigger(input: number) {
    const diff = Math.abs(input - prev);
    const trigger = diff > 0.5;
    prev = input;
    if (input === 0 || prevWasTrigger === true) {
      prevWasTrigger = false;
      return false;
    } else {
      prevWasTrigger = trigger;
      return trigger;
    }
  };
}
