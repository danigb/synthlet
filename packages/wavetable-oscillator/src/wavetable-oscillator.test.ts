import { WavetableOscillator } from "./wavetable-oscillator";

describe("WavetableOscillator", () => {
  it("renders a single plane", () => {
    const osc = WavetableOscillator(10);
    osc.set(new Float32Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]), 10);
    const inputs = {
      frequency: [110],
      baseFrequency: [220],
      morphFrequency: [1],
    };
    const output = new Float32Array(10);
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
  });

  it("renders two planes", () => {
    const osc = WavetableOscillator(10);
    osc.set(
      new Float32Array([
        // first plane
        1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
        // second plane
        101, 102, 103, 104, 105, 106, 107, 108, 109, 110,
        // third plane
        201, 202, 203, 204, 205, 206, 207, 208, 209, 210,
      ]),
      10
    );
    const inputs = {
      frequency: [440],
      baseFrequency: [220],
      morphFrequency: [1],
    };
    const output = new Float32Array(10);
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
    osc.agen(output, inputs);
    expect(output).toMatchSnapshot();
  });
});
