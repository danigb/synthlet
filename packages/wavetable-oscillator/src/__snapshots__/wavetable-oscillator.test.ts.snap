// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WavetableOscillator renders a single plane 1`] = `
Float32Array [
  1,
  1.5,
  2,
  2.5,
  3,
  3.5,
  4,
  4.5,
  5,
  5.5,
]
`;

exports[`WavetableOscillator renders a single plane 2`] = `
Float32Array [
  6,
  6.5,
  7,
  7.5,
  8,
  8.5,
  9,
  9.5,
  10,
  5.5,
]
`;

exports[`WavetableOscillator renders a single plane 3`] = `
Float32Array [
  1,
  1.5,
  2,
  2.5,
  3,
  3.5,
  4,
  4.5,
  5,
  5.5,
]
`;

exports[`WavetableOscillator renders two planes 1`] = `
Float32Array [
  11,
  23,
  35,
  47,
  59,
  61,
  73,
  85,
  97,
  109,
]
`;

exports[`WavetableOscillator renders two planes 2`] = `
Float32Array [
  11,
  123,
  135,
  147,
  159,
  161,
  173,
  185,
  197,
  209,
]
`;

exports[`WavetableOscillator renders two planes 3`] = `
Float32Array [
  111,
  163,
  145,
  127,
  109,
  81,
  63,
  45,
  27,
  9,
]
`;
